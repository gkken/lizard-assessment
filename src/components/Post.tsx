import type { Category } from './MainContainer';
import dayjs from 'dayjs';
import CategoryFlag from './CategoryFlag';
import { removeDuplicateCategories } from 'utils/removeDuplicateCategories';

export interface PostProps {
  publishDate: string;
  summary: string;
  title: string;
  authorName: string;
  authorAvatar: string;
  categories: Category[];
}

const Post = ({
  authorName,
  authorAvatar,
  title,
  publishDate,
  summary,
  categories,
}: PostProps) => {
  return (
    <article className="mb-5 max-w-4xl space-y-2 border-b-2 p-3">
      <div className="flex flex-row space-x-2">
        <img src={authorAvatar} alt="" className="rounded-full" width={24} />
        <h1 className="flex self-center text-sm font-semibold">{authorName}</h1>
      </div>
      <h1 className="text-2xl font-bold">{title}</h1>
      <section className="text-grey-subtext">{summary}</section>
      <footer className="flex flex-row flex-wrap gap-y-1">
        <time className="mr-2 whitespace-nowrap text-sm text-grey-subtext">
          {dayjs(publishDate).format('D MMMM YYYY')}
        </time>
        <div className="flex flex-row flex-wrap gap-y-1">
          {removeDuplicateCategories(categories).map((category) => {
            return (
              <CategoryFlag key={category.id} categoryLabel={category.name} />
            );
          })}
        </div>
      </footer>
    </article>
  );
};

export default Post;
