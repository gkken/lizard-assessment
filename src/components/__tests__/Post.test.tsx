import Post from 'components/Post';
import { render, screen } from '@testing-library/react';

const dummyCategory = [
  {
    id: '5ee1187a-26f3-4819-b710-ccd99efc94df',
    name: 'category1',
  },
  {
    id: 'dc431d44-e26e-4bec-a2bd-a8ba1cd8b95d',
    name: 'category2',
  },
];

const AUTHOR_AVATAR_URL =
  'https://robohash.org/quamnonet.jpg?size=50x50&set=set1';

describe('Post', () => {
  test('Should match snapshot', () => {
    expect(
      render(
        <Post
          authorName={'test authorName'}
          authorAvatar={AUTHOR_AVATAR_URL}
          publishDate={'2020-09-28T15:59:05Z'}
          summary={'test summary'}
          title={'test title'}
          categories={dummyCategory}
        />
      )
    ).toMatchSnapshot();
  });

  test('Passing in all props, should render with appropriate props', () => {
    render(
      <Post
        authorName={'test authorName'}
        authorAvatar={AUTHOR_AVATAR_URL}
        publishDate={'2020-09-28T15:59:05Z'}
        summary={'test summary'}
        title={'test title'}
        categories={dummyCategory}
      />
    );

    const authorName = screen.getByText('test authorName', { selector: 'h1' });
    const title = screen.getByText('test title', { selector: 'h1' });
    const summary = screen.getByText('test summary', { selector: 'section' });
    const authorAvatar = screen.getByRole('img');
    const publishedDate = screen.getByText('28 September 2020', {
      selector: 'time',
    });
    const category1 = screen.getByText('category1');
    const category2 = screen.getByText('category2');

    expect(authorName).toBeInTheDocument();
    expect(title).toBeInTheDocument();
    expect(summary).toBeInTheDocument();
    expect(authorAvatar).toBeInTheDocument();
    expect(authorAvatar).toHaveAttribute('src', AUTHOR_AVATAR_URL);
    expect(publishedDate).toBeInTheDocument();
    expect(category1).toBeInTheDocument();
    expect(category2).toBeInTheDocument();
  });
});
