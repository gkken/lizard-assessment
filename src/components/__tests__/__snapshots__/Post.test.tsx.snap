// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post Should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="mb-5 max-w-4xl space-y-2 border-b-2 p-3"
      >
        <div
          class="flex flex-row space-x-2"
        >
          <img
            alt=""
            class="rounded-full"
            src="https://robohash.org/quamnonet.jpg?size=50x50&set=set1"
            width="24"
          />
          <h1
            class="flex self-center text-sm font-semibold"
          >
            test authorName
          </h1>
        </div>
        <h1
          class="text-2xl font-bold"
        >
          test title
        </h1>
        <section
          class="text-grey-subtext"
        >
          test summary
        </section>
        <footer
          class="flex flex-row flex-wrap gap-y-1"
        >
          <time
            class="mr-2 whitespace-nowrap text-sm text-grey-subtext"
          >
            28 September 2020
          </time>
          <div
            class="flex flex-row flex-wrap gap-y-1"
          >
            <div
              class="mr-2 max-w-max cursor-default whitespace-nowrap rounded-xl bg-grey-light px-2 text-sm text-grey-subtext transition-all duration-300 ease-in hover:bg-grey-dark"
            >
              category1
            </div>
            <div
              class="mr-2 max-w-max cursor-default whitespace-nowrap rounded-xl bg-grey-light px-2 text-sm text-grey-subtext transition-all duration-300 ease-in hover:bg-grey-dark"
            >
              category2
            </div>
          </div>
        </footer>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="mb-5 max-w-4xl space-y-2 border-b-2 p-3"
    >
      <div
        class="flex flex-row space-x-2"
      >
        <img
          alt=""
          class="rounded-full"
          src="https://robohash.org/quamnonet.jpg?size=50x50&set=set1"
          width="24"
        />
        <h1
          class="flex self-center text-sm font-semibold"
        >
          test authorName
        </h1>
      </div>
      <h1
        class="text-2xl font-bold"
      >
        test title
      </h1>
      <section
        class="text-grey-subtext"
      >
        test summary
      </section>
      <footer
        class="flex flex-row flex-wrap gap-y-1"
      >
        <time
          class="mr-2 whitespace-nowrap text-sm text-grey-subtext"
        >
          28 September 2020
        </time>
        <div
          class="flex flex-row flex-wrap gap-y-1"
        >
          <div
            class="mr-2 max-w-max cursor-default whitespace-nowrap rounded-xl bg-grey-light px-2 text-sm text-grey-subtext transition-all duration-300 ease-in hover:bg-grey-dark"
          >
            category1
          </div>
          <div
            class="mr-2 max-w-max cursor-default whitespace-nowrap rounded-xl bg-grey-light px-2 text-sm text-grey-subtext transition-all duration-300 ease-in hover:bg-grey-dark"
          >
            category2
          </div>
        </div>
      </footer>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
